# Workshop2

June 10th

Learn grouping, indexing, filtering

In [1]:
# import packages
import pandas as pd
import numpy as np
import os

In [2]:
os.getcwd() # os.chdir('path_to_dir')

'C:\\Users\\bolim\\Google Drive\\DA_with_python\\python_workshop'

In [3]:
d = pd.read_csv('surveys.csv') # Read in the data

In [4]:
d.head() # Check

record_id  month  day  year  plot_id species_id sex  hindfoot_length  \
0          1      7   16  1977        2         NL   M             32.0   
1          2      7   16  1977        3         NL   M             33.0   
2          3      7   16  1977        2         DM   F             37.0   
3          4      7   16  1977        7         DM   M             36.0   
4          5      7   16  1977        3         DM   M             35.0   

   weight  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN

In [5]:
d.info() # Check

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35549 entries, 0 to 35548
Data columns (total 9 columns):
record_id          35549 non-null int64
month              35549 non-null int64
day                35549 non-null int64
year               35549 non-null int64
plot_id            35549 non-null int64
species_id         34786 non-null object
sex                33038 non-null object
hindfoot_length    31438 non-null float64
weight             32283 non-null float64
dtypes: float64(2), int64(5), object(2)
memory usage: 2.4+ MB


## Getting Summary Statistics

To understand your data, you can use some functions to get quick summary statistics such as the mean value, quartile values, minimum value etc.

In [6]:
d['weight'].describe() # this gives you quick summary stats on the column 'weight'

count    32283.000000
mean        42.672428
std         36.631259
min          4.000000
25%         20.000000
50%         37.000000
75%         48.000000
max        280.000000
Name: weight, dtype: float64

In [7]:
d['weight'].max() # you get the maximum value

280.0

In [9]:
# When you have multiple rows that gives output in one code cell, only the last will print if you don't have a print() statement.
# To print all the rows, make sure you wrap it inside the print(). 
print(d['weight'].min()) 
print(d['weight'].count())
print(d['weight'].std())
print(d['weight'].mean())

4.0
32283
36.63125947458399
42.672428212991356


## 1. Groups

When you 'group' a data, you sort it into groups by the values of the column you specify.

In [13]:
d.groupby(['sex']).size() # Group d into values in column 'sex', and give me the size for each
# Has 15690 female observations, 17348 male observations

sex
F    15690
M    17348
dtype: int64

In [12]:
d.groupby(['sex'])['weight'].mean() # Group d into values in column 'sex'. Then, select the column 'weight', and give the mean value.

sex
F    42.170555
M    42.995379
Name: weight, dtype: float64

Average female weight is 42.17 and average male weight is 42.99

In [14]:
d.groupby(['species_id'])['weight'].mean() # Can do it by species as well

species_id
AB           NaN
AH           NaN
AS           NaN
BA      8.600000
CB           NaN
CM           NaN
CQ           NaN
CS           NaN
CT           NaN
CU           NaN
CV           NaN
DM     43.157864
DO     48.870523
DS    120.130546
DX           NaN
NL    159.245660
OL     31.575258
OT     24.230556
OX     21.000000
PB     31.735943
PC           NaN
PE     21.586508
PF      7.923127
PG           NaN
PH     31.064516
PI     19.250000
PL     19.138889
PM     21.364155
PP     17.173942
PU           NaN
PX     19.000000
RF     13.386667
RM     10.585010
RO     10.250000
RX     15.500000
SA           NaN
SC           NaN
SF     58.878049
SH     73.148936
SO     55.414634
SS     93.500000
ST           NaN
SU           NaN
UL           NaN
UP           NaN
UR           NaN
US           NaN
ZL           NaN
Name: weight, dtype: float64

In [15]:
d.groupby(['sex', 'species_id'])['weight'].mean() # Can group with two columns - this means group by sex and species.

sex  species_id
F    BA              9.161290
     DM             41.609685
     DO             48.531250
     DS            117.749548
     NL            154.282209
     OL             31.065817
     OT             24.830904
     OX             21.000000
     PB             30.210884
     PE             22.822183
     PF              7.974394
     PH             30.850000
     PL             19.312500
     PM             22.125668
     PP             17.180670
     PX             20.000000
     RF             13.684211
     RM             11.102196
     RO             11.000000
     SF             69.000000
     SH             78.756410
     SO             55.322581
     SS             57.000000
M    AH                   NaN
     BA              7.357143
     DM             44.353134
     DO             49.135102
     DS            122.290984
     NL            165.652893
     OL             32.068410
     OT             23.695382
     OX             21.000000
     PB             33.8

Average female weight with species id BA is 9.16

In [16]:
d.groupby(['sex', 'species_id'])['weight'].count()

sex  species_id
F    BA              31
     DM            4440
     DO            1280
     DS            1106
     NL             652
     OL             471
     OT            1029
     OX               2
     PB            1617
     PE             568
     PF             742
     PH              20
     PL              16
     PM             374
     PP            1583
     PX               1
     RF              57
     RM            1184
     RO               4
     SF              16
     SH              78
     SO              31
     SS               1
M    AH               0
     BA              14
     DM            5808
     DO            1621
     DS            1220
     NL             484
     OL             497
     OT            1126
     OX               4
     PB            1188
     PE             689
     PF             802
     PH              11
     PI               7
     PL              19
     PM             498
     PP            1431
     PX               1


`.count()` gives count of values that fall into the category. In the example above, there is 1 observation of a male with species id 'SS'.

## 2. Index

Indexing is slicing the data based on the condition specified. Python uses "zero indexing" - this means that it starts with zero (0). First element is indexed as 0, second element is indexed as 1 and so on.

In [19]:
ex_list = [1, 2, 3, 4, 5] # This is example list

In [20]:
ex_list[0] # Get the element specified in the square bracket.
# This means I want the first element (zero indexing)

1

In [21]:
ex_list[1] # This will give the second element

2

In [23]:
ex_list[4] # This will give the fifth (last) element

5

In [22]:
ex_list[-1] # If you want the last element, you can index from the end, using -1

5

In [24]:
len(ex_list) # Get the length of an object using len(). Length of object "ex_list" is 5.

5

In [25]:
ex_list[5] # will have error, zero indexing

IndexError: list index out of range

FYI: The error says "IndexError: list index out of range". This means that the cause of this error comes from index. (There is nothing wrong with the list) The problem with index is that the value I specified is out of the range. I wanted the sixth element of `ex_list` but `ex_list` has only five elements. That is why it gives an error.

* Index Dataframes

You can index a data frame using `.loc` or `.iloc`. `.iloc` is integer based indexing.

In [26]:
# Index dataframes  
# dataframe[start:stop] (rows)
d[0:3] # first three rows (row 3 is NOT selected)

record_id  month  day  year  plot_id species_id sex  hindfoot_length  \
0          1      7   16  1977        2         NL   M             32.0   
1          2      7   16  1977        3         NL   M             33.0   
2          3      7   16  1977        2         DM   F             37.0   

   weight  
0     NaN  
1     NaN  
2     NaN

In [27]:
d[:3] # This also gives first three rows. Default start is zero, so you don't need to write it

record_id  month  day  year  plot_id species_id sex  hindfoot_length  \
0          1      7   16  1977        2         NL   M             32.0   
1          2      7   16  1977        3         NL   M             33.0   
2          3      7   16  1977        2         DM   F             37.0   

   weight  
0     NaN  
1     NaN  
2     NaN

In [28]:
d.head(3) # That is same as d.head(3) 

record_id  month  day  year  plot_id species_id sex  hindfoot_length  \
0          1      7   16  1977        2         NL   M             32.0   
1          2      7   16  1977        3         NL   M             33.0   
2          3      7   16  1977        2         DM   F             37.0   

   weight  
0     NaN  
1     NaN  
2     NaN

In [29]:
# Subsets of rows and columns with ".loc" or ".iloc"
# dataframe.iloc[row_slicing, column_slicing]
d.iloc[0:5, 0:3] # First five rows, first three rows. Same here - row 5 and column 3 is NOT included.

record_id  month  day
0          1      7   16
1          2      7   16
2          3      7   16
3          4      7   16
4          5      7   16

In [30]:
d.iloc[4:10, 2:4] # rows 5~10, columns 3~4. zero indexing!

day  year
4   16  1977
5   16  1977
6   16  1977
7   16  1977
8   16  1977
9   16  1977

In [31]:
d.loc[0:3, ['species_id', 'plot_id', 'weight']] # With .loc, you can specify with using the column names.

species_id  plot_id  weight
0         NL        2     NaN
1         NL        3     NaN
2         DM        2     NaN
3         DM        7     NaN

In [32]:
d.iloc[0,4] # If you want one value, just need to write dataframe.iloc[row, column]
# This gives first row, fifth column's value.

2

## 3. Filtering

You can filter a dataframe with conditions, which looks like this: ` dataframe[condition_you_want_to_apply]`. The conditions applied will generate a `True` or `False`. You apply the `True` or `False` output on the `dataframe`, and python will apply the `True` and `False` on the dataframe and return ones that says `True`. In other words, it will return the ones that satisfy your condition.

In [36]:
# dataframe[condition_you_want_to_apply]

d_1977 = d[d['year'] == 1977]

Condition applied here is `d['year'] == 1977`. The `year` column value should equal to `1977`. Based on that condition, I will filter the dataframe `d`. I will call that new filtered dataframe as `d_1977`.

* `==` equals to
* `!=` not equals
* `>` greater than
* `<` less than
* `>=` greater than or equal to
* `<=` less than or equal to 

In [37]:
d_1977

record_id  month  day  year  plot_id species_id  sex  hindfoot_length  \
0            1      7   16  1977        2         NL    M             32.0   
1            2      7   16  1977        3         NL    M             33.0   
2            3      7   16  1977        2         DM    F             37.0   
3            4      7   16  1977        7         DM    M             36.0   
4            5      7   16  1977        3         DM    M             35.0   
..         ...    ...  ...   ...      ...        ...  ...              ...   
498        499     12   11  1977       11         DS    F             51.0   
499        500     12   11  1977       21         OT  NaN              NaN   
500        501     12   11  1977       21         OT  NaN              NaN   
501        502     12   11  1977        5         DM    F             36.0   
502        503     12   11  1977       15        NaN  NaN              NaN   

     weight  
0       NaN  
1       NaN  
2       NaN  
3       NaN  
4       NaN  
..      ...  
498   121.0  
499     NaN  
500     NaN  
501    43.0  
502     NaN  

[503 rows x 9 columns]

In [51]:
# Multiple conditions also possible
d_1977_filtered = d[(d['year'] == 1977) & (d['hindfoot_length'] <= 45) & (d['weight'].notna())]

Each condition goes into each parenthesis. There are three conditions here, year should be 1977, hindfoot length should be smaller than or equal to 45, and the weight should not be a null value. (`.notna()` function)

`&` means `and`. `|` means `or`.

In [52]:
d_1977_filtered

record_id  month  day  year  plot_id species_id sex  hindfoot_length  \
62          63      8   19  1977        3         DM   M             35.0   
63          64      8   19  1977        7         DM   M             37.0   
64          65      8   19  1977        4         DM   F             34.0   
65          66      8   19  1977        4         DM   F             35.0   
66          67      8   19  1977        7         DM   M             35.0   
..         ...    ...  ...   ...      ...        ...  ..              ...   
487        488     12   11  1977       12         DM   M             37.0   
489        490     12   11  1977       14         DM   M             37.0   
495        496     12   11  1977       17         DM   F             36.0   
497        498     12   11  1977        3         PF   M             16.0   
501        502     12   11  1977        5         DM   F             36.0   

     weight  
62     40.0  
63     48.0  
64     29.0  
65     46.0  
66     36.0  
..      ...  
487    41.0  
489    44.0  
495    37.0  
497     7.0  
501    43.0  

[227 rows x 9 columns]